File: AndroidStudioProject_Assignment_01/app/src/main/java/com/example/directoryofskilledpeople/DatabaseHelper.java
package com.example.directoryofskilledpeople;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public class DatabaseHelper extends SQLiteOpenHelper {
    //Database Version
    public static final int DATABASE_VERSION = 1;
    //Database Name
     public static final String DATABASE_NAME = "experts_database";
    //Database Table Name
    public static final String TABLE_NAME = "experts_table";
    //Table Columns names
    public static final String ID = "ID";
    public static final String NAME = "NAME";
    public static final String ADDRESS = "ADDRESS";
    public static final String TITLE = "TITLE";
    public static final String PHONE = "PHONE";
    public static final String EMAIL = "EMAIL";
    public static final String NATIONALITY = "NATIONALITY";

    //Database Instance Variable
    private SQLiteDatabase sqLiteDatabase;

    //SQL Queries create Database
    private static final String CREATE_TABLE = "create table " + TABLE_NAME + "("+ID+"TEXT PRIMARY KEY AUTOINCREMENT,"
            +NAME+ "TEXT," +ADDRESS+ "TEXT,"+TITLE + "TEXT, " +PHONE + "TEXT, " + EMAIL + "TEXT, " + NATIONALITY + "TEXT);";



    public DatabaseHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL(CREATE_TABLE);

    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int i, int i1) {
        db.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME);
        onCreate(db);

    }

    //Add Expert Person
    public void addExperts(Skilled_People skilled_People){
        ContentValues contentValues = new ContentValues();
        contentValues.put(DatabaseHelper.NAME, skilled_People.getName());
        contentValues.put(DatabaseHelper.ADDRESS, skilled_People.getAddress());
        contentValues.put(DatabaseHelper.TITLE, skilled_People.getTitle());
        contentValues.put(DatabaseHelper.PHONE, skilled_People.getPhone());
        contentValues.put(DatabaseHelper.EMAIL, skilled_People.getEmail());
        contentValues.put(DatabaseHelper.NATIONALITY, skilled_People.getNationality());
        sqLiteDatabase = this.getWritableDatabase();
        sqLiteDatabase.insert(DatabaseHelper.TABLE_NAME, null, contentValues);
    }

    public List<Skilled_People> getSkilled_people_List(){
        String sql = "select * from " + TABLE_NAME;
        sqLiteDatabase = this.getReadableDatabase();
        List<Skilled_People> store_Skilled_People = new ArrayList<>();
        Cursor cursor = sqLiteDatabase.rawQuery(sql, null);
        if(cursor.moveToFirst()){
            do {
                String id = cursor.getString(0);
                String name = cursor.getString(1);
                String address = cursor.getString(2);
                String title = cursor.getString(3);
                String phone = cursor.getString(4);
                String email = cursor.getString(5);
                String nationality = cursor.getString(6);
                store_Skilled_People.add(new Skilled_People(id,name,address,title,phone,email,nationality));

            }while(cursor.moveToNext());
        }
        cursor.close();
        return store_Skilled_People;
    }
}

File: CRUD/app/src/main/java/com/example/crud/database/DatabaseHelper.kt
/*
    * Created by Yoosuf Sayyid CYX27785 on 21/09/2022
    * This file is part of the project "Notes App" as part of the bachelors degree assignment
    * All rights reserved to Yoosuf Sayyid CYX27785
    *
    * This file is for the database and its functions used in the app
*/

package com.example.crud.database

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import androidx.core.database.getStringOrNull
import com.example.crud.models.TaskListModel

// class initializing the database
class DatabaseHelper (context: Context) : SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {

    // Object that holds information of the DB
    companion object {
        private val DB_NAME = "task"
        private val DB_VERSION = 1
        private val TABLE_NAME = "tasklist"
        private val ID = "id"
        private val TASK_NAME = "taskname"
        private val TASK_DETAILS = "taskdetails"
    }

    // Creating the table
    override fun onCreate(db: SQLiteDatabase?) {
        // creating query
        val CREATE_TABLE = "CREATE TABLE $TABLE_NAME ($ID INTEGER PRIMARY KEY, $TASK_NAME TEXT, $TASK_DETAILS TEXT);"
        // executing query
        db?.execSQL(CREATE_TABLE)
    }

    // executes every time the database is updated
    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        // creating query
        val DROP_TABLE = "DROP TABLE IF EXISTS $TABLE_NAME"
        // executing query
        db?.execSQL(DROP_TABLE)
        // creating table again
        onCreate(db)
    }

    // function to get all the tasks in the database
    // return type will be a list of TaskListModel
    fun getAllTask(): List<TaskListModel> {
        // creating a list of TaskListModel
        val taskList = ArrayList<TaskListModel>()
        // creating / init a readable database
        val db = this.readableDatabase
        // query to get all the data from the table
        val query = "SELECT * FROM $TABLE_NAME"
        // executing the cursor
        // it will get value from the db one by one
        val cursor = db.rawQuery(query, null)

        // if the cursor has data
        // loop through all the cursor until it reaches the end
        if (cursor != null && cursor.moveToFirst()) {
            do {
                // creating a new TaskListModel object
                val task = TaskListModel()
                // get data from the cursor
                task.id = Integer.parseInt(cursor.getString(cursor.getColumnIndexOrThrow(ID)))
                task.name = cursor.getString(cursor.getColumnIndexOrThrow(TASK_NAME))
                task.details = cursor.getString(cursor.getColumnIndexOrThrow(TASK_DETAILS))
                // add the object to the list
                taskList.add(task)
            } while (cursor.moveToNext())
        }
        // close the cursor
        cursor.close()
        // close DB
        db.close()
        // return the list
        return taskList
    }

    // insert operation
    // return type will be a boolean
    fun addTask(task: TaskListModel): Boolean {
        // creating / init a writable database
        val db = this.writableDatabase
        // creating a content value object
        val values = ContentValues()
        // putting data to the content value object
        values.put(TASK_NAME, task.name)
        values.put(TASK_DETAILS, task.details)
        // inserting the data to the table
        // 1 = success
        val success = db.insert(TABLE_NAME, null, values)
        db.close()
        // -1 or 1 is returned depending on the success of the operation
        return (Integer.parseInt("$success") != -1)
    }

    // select data by id
    // gets task when id is passed
    // return type will be a TaskListModel
    fun getTask(id: Int): TaskListModel {
        // creating task object
        val task = TaskListModel()
        // creating / init a readable database
        val db = this.writableDatabase

        // query to get data by id
        val query = "SELECT * FROM $TABLE_NAME WHERE $ID = $id"
        // executing the cursor
        val cursor = db.rawQuery(query, null)

        // if the cursor has data get first
        cursor?.moveToFirst()

        // get data from the cursor
        if (cursor != null && cursor.moveToFirst()) {
            task.id = Integer.parseInt(cursor.getString(cursor.getColumnIndexOrThrow(ID)))
            task.name = cursor.getString(cursor.getColumnIndexOrThrow(TASK_NAME))
            task.details = cursor.getString(cursor.getColumnIndexOrThrow(TASK_DETAILS))
        }

        cursor.close()
        db.close()

        // return the task
        return task
    }

    // delete task by id
    // return type will be a boolean
    fun deleteTask(_id: Int): Boolean {
        val db = this.writableDatabase
        // query is run to delete the task
        // success is returned if the query is successful
        val success = db.delete(TABLE_NAME, "$ID=?", arrayOf(_id.toString())).toLong()
        db.close()
        // -1 or 1 is returned depending on the success of the operation
        return Integer.parseInt("$success") != -1
    }

    // update
    // return type will be a boolean
    fun updateTask(task: TaskListModel): Boolean {
        val db = this.writableDatabase
        // creating a content value object
        val values = ContentValues()
        // putting data to the content value object
        values.put(TASK_NAME, task.name)
        values.put(TASK_DETAILS, task.details)
        // query is run to update the task
        val success = db.update(TABLE_NAME, values, "$ID=?", arrayOf(task.id.toString())).toLong()
        db.close()
        // -1 or 1 is returned depending on the success of the operation
        return Integer.parseInt("$success") != -1
    }
}